vanilla/vanilla
# Add ability to force re-authentication for an action

This update adds the ability to force a user to reauthenticate before they can perform an action. Currently, this is only applied to the "Edit" page on a user's profile, where they can change their email address and username.

### Overview

1. New: `Gdn_Controller::reauth`. This method checks to see if a user needs to reauthenticate. If they do, it performs an internal redirect to send them to the reauthentication page. This method can be dropped into controllers to verify a user has recently been authenticated.
1. New: `ProfileController::authenticate`. This page provides a simple form and a generic explanation of why a user needs to supply their password. It saves the initial POST contents and essentially resubmits the original form, once the user provides a valid password.
1. New: "authenticate" view in the Dashboard application. This is the basic form used by `ProfileController::authenticate`.
1. New: `Gdn_CookieIdentity::getAttribute`. Allows setting an arbitrary value in the JWT payload of the identity cookie.
1. New: `Gdn_CookieIdentity::getAuthTime`. A convenience method for getting the last time the user authenticated.
1. New: `Gdn_CookieIdentity::setJWTPayload`. Sets the full JWT payload to the identity cookie.
1. New: `Gdn_CookieIdentity::setAttribute`. Allows getting a single attribute from the identity cookie's JWT payload.
1. New: `Gdn_CookieIdentity::setAuthTime`. A convenience method for setting the last time the user authenticated.
1. Update: Removed `nbf` claim from the JWT cookie. We never really used it and, ultimately, it didn't make a lot of sense in the context of our cookie.

### Testing

1. Wait until at least 20 minutes have elapsed since your last login.
1. Edit your profile using the /profile/edit page.
1. You should now be prompted to enter your password again.
    - [ ] Enter no password.
    - [ ] Enter the wrong password.
    - [ ] Refresh the page (resubmitting the form) without doing anything.
    - [ ] Enter the valid password.
1. If you've entered a valid password, the original form should reappear and the updated values should be reflected as if you submitted the form, normally.

Closes #5805
